plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java

```diff
@@ -0,0 +1,183 @@
+        this.readTimeoutMillis = readTimeoutMillis;
+    }
+
+    /**
+     * Load all client settings from the given settings.
+     *
+     * Note this will always at least return a client named "default".
+     */
+    static Map<String, S3ClientSettings> load(Settings settings) {
+        Set<String> clientNames = settings.getGroups(PREFIX).keySet();
+        Map<String, S3ClientSettings> clients = new HashMap<>();
+        for (String clientName : clientNames) {
+            clients.put(clientName, getClientSettings(settings, clientName));
+        }
+        if (clients.containsKey("default") == false) {
+            // this won't find any settings under the default client,
+            // but it will pull all the fallback static settings
+            clients.put("default", getClientSettings(settings, "default"));
+        }
+        return Collections.unmodifiableMap(clients);
+    }
+
+    // pkg private for tests
+    /** Parse settings for a single client. */
+    static S3ClientSettings getClientSettings(Settings settings, String clientName) {
+        try (SecureString accessKey = getConfigValue(settings, clientName, ACCESS_KEY_SETTING, S3Repository.Repositories.KEY_SETTING);
+             SecureString secretKey = getConfigValue(settings, clientName, SECRET_KEY_SETTING, S3Repository.Repositories.SECRET_SETTING);
+             SecureString proxyUsername = getConfigValue(settings, clientName, PROXY_USERNAME_SETTING, CLOUD_S3.PROXY_USERNAME_SETTING);
+             SecureString proxyPassword = getConfigValue(settings, clientName, PROXY_PASSWORD_SETTING, CLOUD_S3.PROXY_PASSWORD_SETTING)) {
+            BasicAWSCredentials credentials = null;
+            if (accessKey.length() != 0) {
+                if (secretKey.length() != 0) {
+                    credentials = new BasicAWSCredentials(accessKey.toString(), secretKey.toString());
+                } else {
+                    throw new IllegalArgumentException("Missing secret key for s3 client [" + clientName + "]");
+                }
+            } else if (secretKey.length() != 0) {
+                throw new IllegalArgumentException("Missing access key for s3 client [" + clientName + "]");
+            }
+            return new S3ClientSettings(
+                credentials,
+                getConfigValue(settings, clientName, ENDPOINT_SETTING, S3Repository.Repositories.ENDPOINT_SETTING),
+                getConfigValue(settings, clientName, PROTOCOL_SETTING, S3Repository.Repositories.PROTOCOL_SETTING),
+                getConfigValue(settings, clientName, PROXY_HOST_SETTING, AwsS3Service.CLOUD_S3.PROXY_HOST_SETTING),
+                getConfigValue(settings, clientName, PROXY_PORT_SETTING, AwsS3Service.CLOUD_S3.PROXY_PORT_SETTING),
+                proxyUsername.toString(),
+                proxyPassword.toString(),
+                (int)getConfigValue(settings, clientName, READ_TIMEOUT_SETTING, AwsS3Service.CLOUD_S3.READ_TIMEOUT).millis()
```

Comment:
I assume this is returning a long? Is it okay if the read timeout is -128731 or something like that?